enclave/impl/src/message_bus/mod.rs

```diff
@@ -0,0 +1,753 @@
+        // in order to see when they may be created or deleted.
+        rng.fill_bytes(id.as_mut_slice());
+        while bool::from(id.ct_eq(&Default::default())) {
+            rng.fill_bytes(id.as_mut_slice());
+        }
+
+        // Set the timestamp to our value
+        *marshalled.get_timestamp_mut() = self.current_timestamp;
+        // Set the sender identity to be the authorized identity value.
+        marshalled
+            .get_sender_mut()
+            .copy_from_slice(&query.auth_identity);
+
+        let mut insertion_succeeds = Choice::from(0);
+        let mut message_omap_result_code = 0u32;
+        let recipient_omap_result_code = self.recipients_to_ids.access_and_insert(
+            marshalled.get_recipient(),
+            &Default::default(),
+            &mut rng,
+            |_code, buffer| {
+                let mut found_key = Choice::from(0);
+                let mut found_key_needs_overwrite = Choice::from(0);
+                let mut found_free_space = Choice::from(0);
+                let mut desired_index = u64::MAX;
+
+                // Find a place where we can write a new msg id record for this recipient, if
+                // possible.
+                let chunks: &mut [A8Bytes<U24>] = buffer.as_mut_aligned_chunks();
+                for (idx, chunk) in chunks.iter_mut().enumerate() {
+                    let (msg_id, timestamp): (&mut A8Bytes<U16>, &mut A8Bytes<U8>) = chunk.split();
+
+                    let msg_id_is_zero = msg_id.ct_eq(&Default::default());
+                    let timestamp_expired = (timestamp.as_ne_u64_slice()[0]
+                        + self.message_time_to_live)
+                        .ct_lt(&self.current_timestamp);
+                    let msg_id_is_key = msg_id.ct_eq(&id);
+
+                    // The space is vacant if its id is all zeroes, or the timestamp is expired
+                    found_free_space |= msg_id_is_zero | timestamp_expired;
+                    found_key |= msg_id_is_key;
+                    found_key_needs_overwrite |= msg_id_is_key & timestamp_expired;
+
+                    // If this msg_id matches our key, or it's a free space and we haven't found our
+                    // key already, then we should update the desired_index
+                    // variable.
+                    desired_index.cmov(
+                        msg_id_is_key | (!found_key & found_free_space),
+                        &(idx as u64),
+                    );
+                }
+
+                insertion_succeeds = found_key | found_free_space;
+                message_omap_result_code = self.ids_to_messages.vartime_write_extended(
+                    &id,
+                    &marshalled,
+                    found_key_needs_overwrite,
+                    insertion_succeeds,
+                );
+
+                if message_omap_result_code == OMAP_OVERFLOW {
+                    // Do something
```

Comment:
Not finding this comment very helpful...

Reply:
yeah i still need to work out how the status code determinations and error handling works in this function... this part is still rough